vshard/service_info.lua

```diff
@@ -0,0 +1,91 @@
+--
+-- Every vshard's background service (e.g. rebalancer or discovery)
+-- should have an associated instance of the service_info, which
+-- stores the status and the current activity of the service.
+--
+-- It may be accessed explicitly through the internals of the storage
+-- or the router, which is done for the testing of the background fibers
+-- using luatest. For monitoring purposes the info can be accessed via
+-- vshard.router/storage.info({with_services = true}).
+--
+
+local SERVICE_TEMPLATE = {
+    -- For logging purposes
+    name = nil,
+    -- Intended to show whether everything is all right.
+    status = 'ok',
+    -- Increasing number of status changes.
+    status_idx = 0,
+    -- Intended to show what the service is doing right now.
+    activity = 'unknown',
+    -- Shows whether error was already set on the current iteration.
+    -- Should be dropped on every iteration to false with info:next_iter().
+    is_error_set = false,
+    -- Active error
+    error = '',
+}
+
+local function service_next_iter(service)
+    service.is_error_set = false
+end
+
+local function service_set_status(service, status)
+    service.status_idx = service.status_idx + 1
+    service.status = status
+end
+
+-- Just for consistency with set_status
+local function service_set_activity(service, activity)
+    service.activity = activity
+end
+
+local function service_set_status_ok(service)
+    service_set_status(service, 'ok')
+    service_next_iter(service)
+    service.error = ''
+end
+
+local function service_set_status_error(service, err)
+    -- 'status_idx' is updated even if no error will be saved
+    service_set_status(service, 'error')
```

Comment:
If the error is already set, then we bumped the status index here but didn't update the error. Which might mistakenly be detected like the error occurred again on a new iteration. `service_wait_for_new_error()` would return even though there was no new error. Maybe this function should early-exit if `is_error_set` is true.

Reply:
You're right, let's not update idx.